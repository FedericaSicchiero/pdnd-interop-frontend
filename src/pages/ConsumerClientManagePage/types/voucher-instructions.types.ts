import type { StepperStepComponentProps } from '@/types/common.types'
import type { PublicKeys } from '@/types/key.types'
import type { Purpose } from '@/types/purpose.types'

export type VoucherInstructionsStepProps = StepperStepComponentProps & {
  clientId: string
  clientKeys: PublicKeys
  purpose?: Purpose
  purposeId?: string
}
