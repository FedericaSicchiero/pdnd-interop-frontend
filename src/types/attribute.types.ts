export type BackendAttributeContent = {
  id: string
  name: string
  creationTime: string
  origin: string
  code: string
  description: string
  explicitAttributeVerification: boolean
  verified?: boolean
  verificationDate?: string
}

export type AttributeKey = 'certified' | 'verified' | 'declared'

export type SingleBackendAttribute = {
  single: BackendAttributeContent
}
export type GroupBackendAttribute = {
  group: Array<BackendAttributeContent>
}

export type BackendAttribute = SingleBackendAttribute | GroupBackendAttribute
export type BackendAttributes = Record<AttributeKey, Array<BackendAttribute>>
export type AttributeKind = 'CERTIFIED' | 'VERIFIED' | 'DECLARED'

export type CatalogAttribute = {
  kind: AttributeKind
  creationTime: string
  description: string
  id: string
  name: string
  code?: string
  origin?: 'IPA'
}

export type FrontendAttribute = {
  attributes: Array<CatalogAttribute>
  explicitAttributeVerification: boolean
}

export type FrontendAttributes = Record<AttributeKey, Array<FrontendAttribute>>

export type VerifiedAttributesRenewalType = 'REVOKE_ON_EXPIRATION' | 'AUTOMATIC_RENEWAL'

export type PartyAttributes = Record<AttributeKey, Array<PartyAttribute>>

export type PartyAttribute = {
  id: string
  name: string
  state: 'ACTIVE' | 'REVOKED'
}

export type CertifiedTenantAttribute = {
  id: string
  name: string
  assignmentTimestamp: string
  revocationTimestamp?: string
}

export type VerifiedTenantAttribute = {
  id: string
  name: string
  assignmentTimestamp: string
  revocationTimestamp?: string
  renewal: VerifiedAttributesRenewalType
  verifiedBy: Array<{
    id: string
    verificationDate: string
    expirationDate?: string
    extentionDate?: string
  }>
  revokedBy: Array<{
    id: string
    verificationDate: string
    expirationDate?: string
    extentionDate?: string
    revocationDate: string
  }>
}

export type DeclaredTenantAttribute = {
  id: string
  name: string
  assignmentTimestamp: string
  revocationTimestamp?: string
}
