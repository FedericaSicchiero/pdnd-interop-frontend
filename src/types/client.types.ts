export type ClientPurpose = {
  purposeId: string
  title: string
  eservice: {
    id: string
    name: string
    producer: {
      id: string
      name: string
    }
  }
}

export type Client = {
  id: string
  name: string
  consumer: {
    id: string
    name: string
  }
  purposes: Array<ClientPurpose>
  description: string
  kind: ClientKind
}

export type ClientListingItem = {
  id: string
  name: string
  hasKeys: boolean
}

export type ClientKind = 'CONSUMER' | 'API'
