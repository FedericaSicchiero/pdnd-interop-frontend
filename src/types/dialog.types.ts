import { DialogProps as MUIDialogProps } from '@mui/material'

export type DialogContent = {
  title: string
  description?: string
}

export type DialogDefaultProps = {
  maxWidth?: MUIDialogProps['maxWidth']
}

export type DialogProps =
  | DialogBasicProps
  | DialogAttributeDetailsProps
  | DialogSessionExpiredProps
  | DialogUpdatePurposeDailyCallsProps
  | DialogSetPurposeExpectedApprovalDateProps
  | DialogAddSecurityOperatorsProps
  | DialogAddSecurityOperatorKeyProps
  | DialogRejectAgreementProps

export type DialogAttributeDetailsProps = {
  type: 'showAttributeDetails'
  attribute: { id: string; name: string }
}

export type DialogSessionExpiredProps = {
  type: 'sessionExpired'
}

export type DialogUpdatePurposeDailyCallsProps = {
  type: 'updatePurposeDailyCalls'
  purposeId: string
  dailyCalls?: number
}

export type DialogBasicProps = DialogDefaultProps & {
  type: 'basic'
  title: string
  description?: string
  proceedCallback: () => void
  proceedLabel?: string
  disabled?: boolean
}

export type DialogSetPurposeExpectedApprovalDateProps = {
  type: 'setPurposeExpectedApprovalDate'
  purposeId: string
  versionId: string
  approvalDate?: string
}

export type DialogAddSecurityOperatorsProps = {
  type: 'addSecurityOperator'
  clientId: string
}

export type DialogAddSecurityOperatorKeyProps = {
  type: 'addSecurityOperatorKey'
  clientId: string
}

export type DialogRejectAgreementProps = {
  type: 'rejectAgreement'
  agreementId: string
}
