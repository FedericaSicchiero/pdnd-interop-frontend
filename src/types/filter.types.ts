export type ActiveFilters = Array<FilterOption & { type: FilterFieldType; filterKey: string }>
export type FieldsValues = Record<string, Array<FilterOption> | string>
export type FilterFieldType = 'multiple' | 'single'
export type FiltersParams = Record<string, string | string[] | boolean>

export type FilterField = {
  name: string
  label: string
} & (
  | {
      type: 'multiple'
      options: Array<FilterOption>
      setAutocompleteInput?: React.Dispatch<React.SetStateAction<string>>
    }
  | { type: 'single' }
)

export type FilterFields = Array<FilterField>
export type FilterOption = { label: string; value: string }

export type FiltersHandler = (
  type: FilterFieldType,
  filterKey: string,
  value: string | Array<FilterOption>
) => void
