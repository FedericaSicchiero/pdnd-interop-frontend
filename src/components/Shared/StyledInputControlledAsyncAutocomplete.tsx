import React, { useState } from 'react'
import { Controller } from 'react-hook-form'
import debounce from 'lodash/debounce'
import get from 'lodash/get'
import isEmpty from 'lodash/isEmpty'
import uniqBy from 'lodash/uniqBy'
import { AxiosResponse } from 'axios'
import { Autocomplete, TextField } from '@mui/material'
import { Endpoint } from '../../../types'
import { fetchWithLogs } from '../../lib/api-utils'
import { getFetchOutcome } from '../../lib/error-utils'
import { StyledSpinner } from './StyledSpinner'
import { StyledInputError } from './StyledInputError'
import { InfoMessage } from './InfoMessage'

type StyledInputControlledAsyncAutocompleteProps = {
  label: string
  disabled?: boolean
  infoLabel?: string

  name: string
  defaultValue?: string
  control: any
  rules: any
  errors: any

  placeholder: string
  path: Endpoint
  transformFn: any
  labelKey: string
  multiple?: boolean
}

export function StyledInputControlledAsyncAutocomplete({
  label,
  disabled = false,
  infoLabel,

  name,
  defaultValue,
  control,
  rules,
  errors,

  placeholder,
  path,
  transformFn,
  labelKey,
  multiple = false,
}: StyledInputControlledAsyncAutocompleteProps) {
  const [isOpen, setIsOpen] = useState(false)
  const [options, setOptions] = useState<any>([])
  const [isLoading, setIsLoading] = useState(false)

  const handleSearch = async (e: any) => {
    if (!e) return

    if (!e.target.value) {
      setOptions([])
      return
    }

    setIsLoading(true)

    const searchResponse = await fetchWithLogs({
      path,
      config: { params: { limit: 100, page: 1, search: e.target.value } },
    })

    const outcome = getFetchOutcome(searchResponse)

    if (outcome === 'success') {
      setOptions(transformFn((searchResponse as AxiosResponse).data))
    }

    setIsLoading(false)
  }

  const open = () => {
    setIsOpen(true)
  }

  const close = () => {
    setIsOpen(false)
  }

  const hasFieldError = Boolean(!isEmpty(errors) && !isEmpty(get(errors, name)))

  return (
    <React.Fragment>
      <Controller
        shouldUnregister={true}
        name={name}
        control={control}
        defaultValue={defaultValue}
        rules={rules}
        render={({ field }) => (
          <Autocomplete
            {...field}
            disabled={disabled}
            multiple={multiple}
            open={isOpen}
            onChange={(_, data) => field.onChange(data)}
            onInputChange={debounce(handleSearch, 100)}
            onOpen={open}
            onClose={close}
            getOptionLabel={(option: any) => (option ? option[labelKey] : '')}
            isOptionEqualToValue={(option: any, value: any) => option[labelKey] === value[labelKey]}
            filterOptions={(options) => uniqBy(options, (o) => o[labelKey].toLowerCase())}
            options={options}
            loading={isLoading}
            loadingText="Stiamo cercando..."
            noOptionsText="Nessun risultato trovato"
            renderInput={(params) => {
              return (
                <TextField
                  label={label}
                  {...params}
                  placeholder={placeholder}
                  variant="standard"
                  InputProps={{
                    ...params.InputProps,
                    endAdornment: (
                      <React.Fragment>
                        {isLoading ? <StyledSpinner color="primary" size={20} /> : null}
                        {params.InputProps.endAdornment}
                      </React.Fragment>
                    ),
                  }}
                  error={hasFieldError}
                />
              )
            }}
          />
        )}
      />

      {hasFieldError && <StyledInputError error={errors[name]} />}
      {infoLabel && <InfoMessage label={infoLabel} />}
    </React.Fragment>
  )
}
