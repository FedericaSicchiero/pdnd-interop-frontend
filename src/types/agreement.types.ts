import {
  AttributeKey,
  CertifiedTenantAttribute,
  DeclaredTenantAttribute,
  VerifiedTenantAttribute,
} from './attribute.types'
import { DocumentRead } from './common.types'
import { EServiceState } from './eservice.types'

export type AgreementState =
  | 'ACTIVE'
  | 'SUSPENDED'
  | 'PENDING'
  | 'ARCHIVED'
  | 'DRAFT'
  | 'REJECTED'
  | 'MISSING_CERTIFIED_ATTRIBUTES'

type AgreementProducer = {
  name: string
  id: string
}
type AgreementConsumer = {
  name: string
  id: string
  attributes: Array<
    Record<
      AttributeKey,
      DeclaredTenantAttribute | CertifiedTenantAttribute | VerifiedTenantAttribute
    >
  >
}

type AgreementEService = {
  name: string
  id: string
  descriptorId: string
  version: string
  state: EServiceState
  activeDescriptor?: AgreementEService
}

export type AgreementAttribute = {
  id: string
  description: string
  name: string
  creationTime: string
}

export type AgreementListingItem = {
  id: string
  state: AgreementState
  consumer: {
    id: string
    name: string
  }
  eservice: {
    id: string
    name: string
    producer: {
      id: string
      name: string
    }
  }
  descriptor: {
    version: number
  }
  suspendedByConsumer: boolean
  suspendedByProducer: boolean
  suspendedByPlatform: boolean
  canBeUpgraded: boolean
}

export type AgreementSummary = {
  id: string
  state: AgreementState
  eservice: AgreementEService
  descriptorId: string
  consumer: AgreementConsumer
  producer: AgreementProducer
  verifiedAttributes: Array<AgreementAttribute>
  certifiedAttributes: Array<AgreementAttribute>
  declaredAttributes: Array<AgreementAttribute>
  suspendedByProducer?: boolean
  suspendedByConsumer?: boolean
  suspendedByPlatform?: boolean
  consumerNotes?: string
  consumerDocuments: Array<DocumentRead>
  createdAt: string
  updatedAt?: string
  rejectionReason?: string
  isContractPresent: boolean
}
