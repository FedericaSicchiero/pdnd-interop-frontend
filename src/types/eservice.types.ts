import { AgreementState } from './agreement.types'
import { BackendAttribute, BackendAttributes } from './attribute.types'
import { DocumentRead } from './common.types'

type EServiceReadProducerType = {
  id: string
  name: string
}

export type EServiceTechnologyType = 'REST' | 'SOAP'
export type EServiceState = 'PUBLISHED' | 'DRAFT' | 'SUSPENDED' | 'ARCHIVED' | 'DEPRECATED'

export type EServiceDocumentKind = 'INTERFACE' | 'DOCUMENT'

export type EServiceReadType = {
  id: string
  producer: EServiceReadProducerType
  name: string
  description: string
  technology: EServiceTechnologyType
  attributes: BackendAttributes
  descriptors: Array<EServiceDescriptorRead>
  state?: EServiceState
  viewingDescriptor?: EServiceDescriptorRead // TEMP REFACTOR : this is added by the client
}

/** @deprecated TO BE REMOVED */
export type EServiceFlatten = {
  name: string
  description: string
  id: string
  producerId: string
  producerName: string
  descriptorId?: string
  state?: EServiceState
  version?: string
  agreement?: {
    id: string
    state: AgreementState
  }
  certifiedAttributes: Array<BackendAttribute>
}

export type EServiceCatalog = {
  id: string
  name: string
  description: string
  producer: {
    id: string
    name: string
  }
  agreement: {
    id: string
    state: AgreementState
  }
  isMine: true
  canSubscribe: true
  activeDescriptor: {
    id: string
    state: EServiceState
    version: string
  }
}

export type EServiceDescriptorRead = {
  id: string
  state: EServiceState
  docs: Array<DocumentRead>
  interface?: DocumentRead
  version: string
  voucherLifespan: number
  description: string
  audience: Array<string>
  dailyCallsPerConsumer: number
  dailyCallsTotal: number
  agreementApprovalPolicy: 'MANUAL' | 'AUTOMATIC'
}
