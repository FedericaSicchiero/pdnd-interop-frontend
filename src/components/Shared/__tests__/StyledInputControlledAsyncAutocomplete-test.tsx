import React from 'react'
import renderer from 'react-test-renderer'
import noop from 'lodash/noop'
import { ApiEndpointKey, Party } from '../../../../types'
import { StyledInputControlledAsyncAutocomplete } from '../StyledInputControlledAsyncAutocomplete'

describe('Snapshot', () => {
  it.only('matches', () => {
    function AsyncAutocomplete() {
      const props = {
        label: 'Snapshot autocomplete',
        name: 'autocomplete',
        defaultValue: null,
        placeholder: '...',
        onChange: noop,
        path: { endpoint: 'ONBOARDING_GET_AVAILABLE_PARTIES' as ApiEndpointKey },
        transformKey: 'institutions',
        transformFn: (data: Array<Party>) => data,
        getOptionLabel: (option: Party) => (option ? option.description : ''),
        isOptionEqualToValue: (option: Party, value: Party) => option.id === value.id,
      }
      return <StyledInputControlledAsyncAutocomplete {...props} />
    }

    const component = renderer.create(<AsyncAutocomplete />)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
