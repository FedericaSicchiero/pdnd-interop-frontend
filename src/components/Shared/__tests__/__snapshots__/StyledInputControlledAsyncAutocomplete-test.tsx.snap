// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches 1`] = `
<div
  className="MuiBox-root css-z727b5"
>
  <div
    aria-expanded={false}
    aria-owns={null}
    className="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1kpdewa-MuiAutocomplete-root"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="combobox"
  >
    <div
      className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={false}
        id="undefined-label"
      >
        Snapshot autocomplete
      </label>
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-activedescendant={null}
          aria-autocomplete="list"
          aria-controls={null}
          aria-invalid={false}
          autoCapitalize="none"
          autoComplete="off"
          autoFocus={false}
          className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          placeholder="..."
          required={false}
          spellCheck="false"
          type="text"
          value=""
        />
        <div
          className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
        >
          <button
            aria-label="Open"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            title="Open"
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-dhh0px"
          >
            <span>
              Snapshot autocomplete
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
